From 7088fa331408badea63a2bc00f60cb6c3d182d29 Mon Sep 17 00:00:00 2001
From: David Benjamin <davidben@google.com>
Date: Wed, 27 Nov 2019 17:02:13 -0500
Subject: [PATCH 01/15] Don't call ERR_error_string in keymaster

ERR_error_string(nullptr) puts the error in a global buffer, which is
not thread-safe. Use ERR_error_string_n instead.

Test: mm
Change-Id: I55b3443b26e9bd7bdfce7f37290520dd6c625954
---
 km_openssl/block_cipher_operation.cpp | 4 +++-
 km_openssl/openssl_err.cpp            | 4 +++-
 2 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/km_openssl/block_cipher_operation.cpp b/km_openssl/block_cipher_operation.cpp
index 20bd2a6..7ba3707 100644
--- a/km_openssl/block_cipher_operation.cpp
+++ b/km_openssl/block_cipher_operation.cpp
@@ -197,7 +197,9 @@ keymaster_error_t BlockCipherEvpOperation::Finish(const AuthorizationSet& additi
     int output_written = -1;
     if (!EVP_CipherFinal_ex(&ctx_, output->peek_write(), &output_written)) {
         if (tag_length_ > 0) return KM_ERROR_VERIFICATION_FAILED;
-        LOG_E("Error encrypting final block: %s", ERR_error_string(ERR_peek_last_error(), nullptr));
+        char buf[128];
+        ERR_error_string_n(ERR_peek_last_error(), buf, sizeof(buf));
+        LOG_E("Error encrypting final block: %s", buf);
         return TranslateLastOpenSslError();
     }
 
diff --git a/km_openssl/openssl_err.cpp b/km_openssl/openssl_err.cpp
index 1391f40..d94897e 100644
--- a/km_openssl/openssl_err.cpp
+++ b/km_openssl/openssl_err.cpp
@@ -44,7 +44,9 @@ keymaster_error_t TranslateLastOpenSslError(bool log_message) {
     unsigned long error = ERR_peek_last_error();
 
     if (log_message) {
-        LOG_D("%s", ERR_error_string(error, nullptr));
+        char buf[128];
+        ERR_error_string_n(error, buf, sizeof(buf));
+        LOG_D("%s", buf);
     }
 
     int reason = ERR_GET_REASON(error);
-- 
2.17.1

